### PR TITLE
Implement external metric in HPA

### PR DESCRIPTION
**Pull request by MaciekPytel**
This implement the changes to HPA introduced in https://github.com/kubernetes/kubernetes/pull/60096

### CODE REVIEW

**Comment by MaciekPytel** [comment]
/hold
This is based on https://github.com/kubernetes/kubernetes/pull/60096, https://github.com/kubernetes/kubernetes/pull/60079 and https://github.com/kubernetes/kubernetes/pull/60220. It must not be merged before all those.

Note: this PR is just implementation for implementation in HPA (last 2 commits). Everything else is just the result of including commits from PRs listed above.

/release-note-none
/sig-autoscaling

File: pkg/controller/podautoscaler/horizontal.go [hunk 1]
```diff
@@ -299,6 +299,45 @@ func (a *HorizontalController) computeReplicasForMetrics(hpa *autoscalingv2.Hori
...
+					},
+				}
+			} else if metricSpec.External.TargetValue != nil {
+				replicaCountProposal, utilizationProposal, timestampProposal, err = a.replicaCalc.GetExternalMetricReplicas(currentReplicas, metricSpec.External.TargetValue.MilliValue(), metricSpec.External.MetricName, hpa.Namespace, metricSpec.External.MetricSelector)
+				if err != nil {
+					a.eventRecorder.Event(hpa, v1.EventTypeWarning, "FailedGetExternalMetric", err.Error())
+					setCondition(hpa, autoscalingv2.ScalingActive, v1.ConditionFalse, "FailedGetExternalMetric", "the HPA was unable to compute the replica count: %v", err)
+					return 0, "", nil, time.Time{}, fmt.Errorf("failed to get external metric %s: %v", metricSpec.External.MetricName, err)
+				}
+				metricNameProposal = fmt.Sprintf("external metric %s", metricSpec.External.MetricName)
```
Comment (DirectXMan12) [inline pkg/controller/podautoscaler/horizontal.go]: should also list label selector here.

File: pkg/controller/podautoscaler/horizontal_test.go [hunk 2]
```diff
@@ -466,8 +467,8 @@ func (tc *testCase) prepareTestClient(t *testing.T) (*fake.Clientset, *metricsfa
 			metrics := &cmapi.MetricValueList{}
 
 			// multiple objects
-			assert.Equal(t, "pods", getForAction.GetResource().Resource, "the type of object that we requested multiple metrics for should have been pods")
-			assert.Equal(t, "qps", getForAction.GetMetricName(), "the metric name requested should have been qps, as specified in the metric spec")
+			assert.Equal(t, "pods", getForAction.GetResource().Resource, fmt.Sprintf("got: %v, want: pods", getForAction.GetResource().Resource))
```
Comment (DirectXMan12) [inline pkg/controller/podautoscaler/horizontal_test.go]: no need for the extra `fmt.Sprintf`, IIRC -- the terminating arguments here are themselves a format string.

Comment (DirectXMan12) [reply pkg/controller/podautoscaler/horizontal_test.go]: also, what's with this change -- we already have a "want" and "got" as part of the assert.  The message is just providing more context, and the way you've changed it provides less.

File: pkg/controller/podautoscaler/horizontal_test.go [hunk 3]
```diff
@@ -526,6 +527,31 @@ func (tc *testCase) prepareTestClient(t *testing.T) (*fake.Clientset, *metricsfa
...
+		defer tc.Unlock()
+
+		getForAction, wasGetFor := action.(emfake.GetForAction)
+		if !wasGetFor {
+			return true, nil, fmt.Errorf("expected a get-for action, got %v instead", action)
+		}
+
+		metrics := &emapi.ExternalMetricValueList{}
+
+		assert.Equal(t, "qps", getForAction.GetMetricName(), fmt.Sprintf("got: %v, want: qps", getForAction.GetMetricName()))
```
Comment (DirectXMan12) [inline pkg/controller/podautoscaler/horizontal_test.go]: the message here should be more expressive (see above)

File: pkg/controller/podautoscaler/replica_calculator.go [hunk 4]
```diff
@@ -286,3 +313,30 @@ func (c *ReplicaCalculator) GetObjectMetricReplicas(currentReplicas int32, targe
...
+		utilization = utilization + val
+	}
+
+	replicaCount = currentReplicas
+	usageRatio := float64(utilization) / (float64(targetUtilizationPerPod) * float64(replicaCount))
+	if math.Abs(1.0-usageRatio) > c.tolerance {
+		// update number of replicas if the change is large enough
+		replicaCount = int32(math.Ceil(float64(utilization) / float64(targetUtilizationPerPod)))
+	}
+	utilization = int64(math.Ceil(float64(utilization) / float64(currentReplicas)))
```
Comment (DirectXMan12) [inline pkg/controller/podautoscaler/replica_calculator.go]: This needs to follow the same logic as `pods`, so it should factor in unready pods.

File: pkg/controller/podautoscaler/replica_calculator_test.go [hunk 5]
```diff
@@ -292,7 +326,7 @@ func (tc *replicaCalcTestCase) runTest(t *testing.T) {
 		require.NoError(t, err, "there should not have been an error calculating the replica count")
 		assert.Equal(t, tc.expectedReplicas, outReplicas, "replicas should be as expected")
 		assert.Equal(t, tc.metric.expectedUtilization, outUtilization, "utilization should be as expected")
-		assert.True(t, tc.timestamp.Equal(outTimestamp), "timestamp should be as expected")
+		assert.True(t, tc.timestamp.Equal(outTimestamp), fmt.Sprintf("got: %v, want: %v", outTimestamp, tc.timestamp))
```
Comment (DirectXMan12) [inline pkg/controller/podautoscaler/replica_calculator_test.go]: again, please don't make the message redundant and less expressive

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
comments inline

File: pkg/controller/podautoscaler/replica_calculator.go (hunk 4 above)
Comment (MaciekPytel) [reply pkg/controller/podautoscaler/replica_calculator.go]: The logic for handling unready pods in `pods` is about how to account for such pods when calculating total value of metric. In this case we get a single metric value and just calculate the desired number of pods. I don't see how logic in `pods` could apply here?

**Comment by MaciekPytel** [comment]
@DirectXMan12 fixed your comments (except the one about unready metrics, which I replied to) and rebased.

File: pkg/controller/podautoscaler/replica_calculator_test.go [hunk 6]
```diff
@@ -425,6 +462,36 @@ func TestReplicaCalcScaleUpCMObject(t *testing.T) {
...
+func TestReplicaCalcScaleUpPerPodCMExternal(t *testing.T) {
+	tc := replicaCalcTestCase{
+		currentReplicas:  3,
+		expectedReplicas: 4,
+		metric: &metricInfo{
+			name:                    "qps",
+			levels:                  []int64{8600},
+			perPodTargetUtilization: 2150,
+			expectedUtilization:     2867,
+			selector:                &metav1.LabelSelector{},
```
Comment (DirectXMan12) [inline pkg/controller/podautoscaler/replica_calculator_test.go]: can we get a test that uses LabelSelector and verifies that it's passed through correctly?

File: pkg/controller/podautoscaler/replica_calculator.go (hunk 4 above)
Comment (DirectXMan12) [reply pkg/controller/podautoscaler/replica_calculator.go]: ack

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
minor testing nit inline.  otherwise looks good

**Review by DirectXMan12 — APPROVED** [review]
you should be good once tests pass

File: pkg/controller/podautoscaler/horizontal.go [hunk 7]
```diff
@@ -299,6 +299,45 @@ func (a *HorizontalController) computeReplicasForMetrics(hpa *autoscalingv2.Hori
 					},
 				}
 			}
+		case autoscalingv2.ExternalMetricSourceType:
+			if metricSpec.External.TargetAverageValue != nil {
```
Comment (mwielgus) [inline pkg/controller/podautoscaler/horizontal.go]: Has it been verified manually E2e?

File: pkg/controller/podautoscaler/horizontal.go [hunk 8]
```diff
@@ -299,6 +299,45 @@ func (a *HorizontalController) computeReplicasForMetrics(hpa *autoscalingv2.Hori
...
+				metricNameProposal = fmt.Sprintf("external metric %s(%+v)", metricSpec.External.MetricName, metricSpec.External.MetricSelector)
+				statuses[i] = autoscalingv2.MetricStatus{
+					Type: autoscalingv2.ExternalMetricSourceType,
+					External: &autoscalingv2.ExternalMetricStatus{
+						MetricSelector:      metricSpec.External.MetricSelector,
+						MetricName:          metricSpec.External.MetricName,
+						CurrentAverageValue: resource.NewMilliQuantity(utilizationProposal, resource.DecimalSI),
+					},
+				}
+			} else if metricSpec.External.TargetValue != nil {
```
Comment (mwielgus) [inline pkg/controller/podautoscaler/horizontal.go]: Has it been validated manually e2e?

File: pkg/controller/podautoscaler/horizontal.go (hunk 7 above)
Comment (MaciekPytel) [reply pkg/controller/podautoscaler/horizontal.go]: Yes

File: pkg/controller/podautoscaler/horizontal.go (hunk 8 above)
Comment (MaciekPytel) [reply pkg/controller/podautoscaler/horizontal.go]: Yes

File: pkg/controller/podautoscaler/horizontal.go [hunk 9]
```diff
@@ -299,6 +299,45 @@ func (a *HorizontalController) computeReplicasForMetrics(hpa *autoscalingv2.Hori
...
+				metricNameProposal = fmt.Sprintf("external metric %s(%+v)", metricSpec.External.MetricName, metricSpec.External.MetricSelector)
+				statuses[i] = autoscalingv2.MetricStatus{
+					Type: autoscalingv2.ExternalMetricSourceType,
+					External: &autoscalingv2.ExternalMetricStatus{
+						MetricSelector: metricSpec.External.MetricSelector,
+						MetricName:     metricSpec.External.MetricName,
+						CurrentValue:   *resource.NewMilliQuantity(utilizationProposal, resource.DecimalSI),
+					},
+				}
+			} else {
```
Comment (mwielgus) [inline pkg/controller/podautoscaler/horizontal.go]: Do we have a check for this in validate logic?

Comment (MaciekPytel) [reply pkg/controller/podautoscaler/horizontal.go]: Yes, https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/autoscaling/validation/validation.go#L200

File: pkg/controller/podautoscaler/metrics/interfaces.go [hunk 10]
```diff
@@ -42,4 +42,8 @@ type MetricsClient interface {
 	// GetObjectMetric gets the given metric (and an associated timestamp) for the given
 	// object in the given namespace
 	GetObjectMetric(metricName string, namespace string, objectRef *autoscaling.CrossVersionObjectReference) (int64, time.Time, error)
+
+	// GetExternalMetric gets all the values of a given external metric
+	// that match the specified selector.
+	GetExternalMetric(metricName string, namespace string, selector labels.Selector) ([]int64, time.Time, error)
```
Comment (mwielgus) [inline pkg/controller/podautoscaler/metrics/interfaces.go]: Lets include Millis in the name of the metrics. It is not obvious from the method name that it returns the value in millis. Here and elsewhere. Can be next PR.

Comment (MaciekPytel) [reply pkg/controller/podautoscaler/metrics/interfaces.go]: Ok, I'll do it in a PR later on (it's just a naming change, no functional impact).

**Pull request merged**